#154792 - CupertinoActionSheetAction cursor doesn't change to clickable on desktop

srivats22 opened the pull request:
This PR is for issue #154792 
The Video recording shows the before and after
The first button with the text "Go" uses the default cursor
The second button with the text "With Cursor" uses a custom cursor that has been passed in the parameters
https://github.com/user-attachments/assets/e82ecd42-42b1-42c9-aa30-a6f3daddb436

Here is the code for the second button
```dart
CupertinoButton(
    onPressed: () {
        showCupertinoModalPopup<void>(
            context: context,
            builder: (BuildContext context) {
                return CupertinoActionSheet(
                    title: const Text('The title'),
                    message: const Text('Message'),
                    actions: <Widget>[
                        CupertinoActionSheetAction(
                            cursor: SystemMouseCursors.forbidden,
                            child: const Text('One'),
                            onPressed: () {},
                        ),
                    ],
                );
            },
        );
    },
    child: const Text('With Cursor'),
),
```

Fixes https://github.com/flutter/flutter/issues/154792
Part of https://github.com/flutter/flutter/issues/58192

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

srivats22 committed:
#154792 - CupertinoActionSheet cursor

srivats22 committed:
Merge pull request #1 from srivats22/i154792

#154792 - CupertinoActionSheet cursor

srivats22 committed:
Merge branch 'master' into master

srivats22 committed:
Merge branch 'master' into master

victorsanni commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1215,6 +1215,7 @@ class CupertinoActionSheetAction extends StatefulWidget {
     required this.onPressed,
     this.isDefaultAction = false,
     this.isDestructiveAction = false,
+    this.cursor = MouseCursor.defer,
```
Wondering if we can use `mouseCursor` instead of `cursor` to be consistent with other material/cupertino APIs that offer a configurable mouse cursor. For example, refer to uses of [mouseCursor](https://github.com/search?q=repo%3Aflutter%2Fflutter%20this.mouseCursor&type=code) vs. that of [cursor](https://github.com/search?q=repo%3Aflutter%2Fflutter+this.cursor&type=code).

srivats22 replied:
sure I can update it to use mouseCursor instead of cursor for consistency

srivats22 replied:
making as resolved taken care in most recent PR

victorsanni commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1312,7 +1318,7 @@ class _CupertinoActionSheetActionState extends State<CupertinoActionSheetAction>
 
     return MouseRegion(
-      cursor: kIsWeb ? SystemMouseCursors.click : MouseCursor.defer,
+      cursor: kIsWeb ? SystemMouseCursors.click : widget.cursor,
```
Can the cursor be configured to accept `WidgetStateMouseCursor`s? It would be nice to have that functionality added right off the bat.

srivats22 replied:
would like some help with this... not sure how to exactly approach it

would it be similar to this pr: https://github.com/flutter/flutter/pull/151788/files

srivats22 replied:
Is this better:
```dart
final WidgetStateProperty<MouseCursor> effectiveMouseCursor =
    WidgetStateProperty.resolveWith<MouseCursor>((Set<WidgetState> states) {
      return WidgetStateProperty.resolveAs<MouseCursor?>(widget.mouseCursor, states)
          ?? (kIsWeb && !states.contains(WidgetState.disabled)
              ? SystemMouseCursors.click
              : SystemMouseCursors.basic
          );
    });

    return MouseRegion(
      cursor: effectiveMouseCursor.resolve(<WidgetState>{}),
      child: MetaData(
        metaData: this,
        behavior: HitTestBehavior.opaque,
        child: ConstrainedBox(
          constraints: const BoxConstraints(
            minHeight: _kActionSheetButtonMinHeight,
          ),
          child: Semantics(
            button: true,
            onTap: widget.onPressed,
            child: Padding(
              padding: EdgeInsets.fromLTRB(
                _kActionSheetButtonHorizontalPadding,
                verticalPadding,
                _kActionSheetButtonHorizontalPadding,
                verticalPadding,
              ),
              child: DefaultTextStyle(
                style: style,
                textAlign: TextAlign.center,
                child: Center(child: widget.child),
              ),
            ),
          ),
        ),
      ),
    );

victorsanni replied:
Only the `resolveAs` block is needed. The wrapping with `resolveWith` was done in the other PR because the  `buildToggleable` method required its mouse cursor to be a `WidgetStateProperty<MouseCursor>` instead of just a `MouseCursor`. But that is not the case here with `MouseRegion`.

srivats22 replied:
```dart
final MouseCursor effectiveMouseCursor =
        WidgetStateProperty.resolveAs<MouseCursor?>(widget.mouseCursor, states)
            ?? (kIsWeb && !states.contains(WidgetState.disabled)
            ? SystemMouseCursors.click
            : SystemMouseCursors.basic
        );

    return MouseRegion(
      cursor: effectiveMouseCursor,
```
just a snippet of the code since everything else is pretty much the same

victorsanni replied:
I'm not sure CupertinoActionSheetAction has a `states` getter. So checking that an action is not disabled instead should be sufficient.

I recommend playing around with it locally and writing tests until they pass. It's easier to review commits vs comment snippets here and I personally find that playing with a system on my own vastly improves my understanding of it.

srivats22 replied:
makes sense, let me play around more and see what best way to do it and will update the pr accordingly.

Thanks

srivats22 replied:
marking as resolved taken care in recent pr

victorsanni commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1234,6 +1235,11 @@ class CupertinoActionSheetAction extends StatefulWidget {
 
+  /// The cursor that will be shown when hovering over the button.
+  ///
+  /// If not passed will default to [MouseCursor.defer] and on web will default to [SystemMouseCursors.click].
```
```suggestion
  /// Defaults to [MouseCursor.defer], except on web, where this defaults to [SystemMouseCursors.click].
```
'Default' makes 'If not passed' redundant.

srivats22 replied:
marking as resolved taken care in most recent PR

srivats22 committed:
Updates based on review comments

1. renames cursor to mouseCursor for consistency
2. mouseCursor now uses WidgetStateProperty so if mouseCursor is not provided on web when hovering over the button SystemMouseCursors.click is rendered and on other platforms will defer to MouseCursor.defer

srivats22 committed:
Merge pull request #2 from srivats22/i154792

Updates based on review comments

victorsanni commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1311,8 +1332,19 @@ class _CupertinoActionSheetActionState extends State<CupertinoActionSheetAction>
     final double verticalPadding = _kActionSheetButtonVerticalPaddingBase
         + fontSize * _kActionSheetButtonVerticalPaddingFactor;
 
+    final Set<WidgetState> states = <WidgetState>{
```
Detecting when a `WidgetState` is active is our responsibility. Typically it's done by having different flags that are activated depending on some metric. For example, if we want to track the `scrolledUnder` state, we would have to get the ScrollMetrics of the scroll view and do the math to calculate when the widget has been scrolled under (and when it has, add `WidgetState.scrolledUnder` to `states`.

It's more complex logic than anticipated, so it's okay to leave it out of this PR and handle it subsequently on a case-by-case basis (for each relevant state).

srivats22 replied:
oh ok, so do you want me to rollback to the original pr?

srivats22 replied:
marking as resolved taken care in recent commit

srivats22 committed:
Reverting to original PR since using WidgetState turned out to be more complicated and will be taken care later

srivats22 committed:
Merge pull request #3 from srivats22/i154792

Reverting to original PR since using WidgetState turned out to be mor…

victorsanni commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1311,8 +1317,12 @@ class _CupertinoActionSheetActionState extends State<CupertinoActionSheetAction>
     final double verticalPadding = _kActionSheetButtonVerticalPaddingBase
         + fontSize * _kActionSheetButtonVerticalPaddingFactor;
 
+    final Set<WidgetState> states = <WidgetState>{
+      WidgetState.hovered
+    };
+
```
```suggestion
```

victorsanni commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1311,8 +1317,12 @@ class _CupertinoActionSheetActionState extends State<CupertinoActionSheetAction>
+
     return MouseRegion(
-      cursor: kIsWeb ? SystemMouseCursors.click : MouseCursor.defer,
+      cursor: kIsWeb ? SystemMouseCursors.click : widget.mouseCursor,
```
What if a user wants to configure a mouse cursor on web (instead of just click)? The user provided property should always take priority.

srivats22 replied:
want a suggestion if I can do the following:
`final MouseCursor webCursor = (kIsWeb && widget.mouseCursor != MouseCursor.defer) ? 
    widget.mouseCursor : SystemMouseCursors.click;`

Since mouseCursor is defaulted to MouseCursor.defer when its web and the mouseCursor is not equal to MouseCursor.defer then we use the cursor passed by the user else we default to SystemMouseCursor.click

victorsanni replied:
Instead of making `mouseCursor` default to `MouseCursor.defer` in the constructor, you can make it nullable. Then when setting the `MouseRegion`s cursor, have `mouseCursor` null coalesce to either `.click` on web or `.defer` otherwise.

srivats22 committed:
removed unnecessary code

Co-authored-by: Victor Sanni <victorsanniay@gmail.com>

victorsanni commented on packages/flutter/test/cupertino/action_sheet_test.dart:
```diff
@@ -1956,6 +1956,41 @@ void main() {
+    const SystemMouseCursor customCursor = SystemMouseCursors.click;
+
+    await tester.pumpWidget(
+        createAppWithButtonThatLaunchesActionSheet(
+            CupertinoActionSheet(
```
Outdent by two spaces.

victorsanni commented on packages/flutter/test/cupertino/action_sheet_test.dart:
```diff
@@ -1956,6 +1956,41 @@ void main() {
+                  child: const Text('One'),
+                ),
+              ],
+            )
+        )
```
Use trailing commas.

victorsanni commented on packages/flutter/test/cupertino/action_sheet_test.dart:
```diff
@@ -1956,6 +1956,41 @@ void main() {
   });
 
+  testWidgets('Hovering over Cupertino action sheet action updates to the passed cursor on desktop', (WidgetTester tester) async {
+    const SystemMouseCursor customCursor = SystemMouseCursors.click;
```
Consider using another cursor to make the logic in the test independent of the "clickable on web" behavior.

victorsanni replied:
Still uses `.click`.

victorsanni replied:
This change, although small, is important because if this PR is regressed, and this test is run on web, it would still pass.

srivats22 replied:
oh so for web use a different one like maybe grab and for non web use something like click or forbidden?

victorsanni replied:
There doesn't need to be different cursors per platform. Just one cursor that is definitely not a default.

Correction: The test wouldn't pass either way because we're adding a property to a public API. It would only pass if the changes are regressed in a way that the property is no longer respected.

srivats22 replied:
understood.

So how would I handle it? Still not able to figure that portion out

victorsanni replied:
Replace `.click` with `.grab` or `.forbidden` etc.

srivats22 replied:
oh ok will update with that... 

apart from this change, others all look good or any changes still required?

victorsanni commented on packages/flutter/test/cupertino/action_sheet_test.dart:
```diff
@@ -1956,6 +1956,41 @@ void main() {
+    await tester.pumpAndSettle();
+    expect(
+      RendererBinding.instance.mouseTracker.debugDeviceActiveCursor(1),
+      kIsWeb ? SystemMouseCursors.click : customCursor,
```
If the test is checking that the mouse cursor provided is always respected, then it shouldn't differ in behavior on web vs on other platforms.

srivats22 replied:
will change it so ensure the mouseCursor is always respected instead of checking platform

srivats22 committed:
Changes related to comment discussion

srivats22 committed:
Merge pull request #4 from srivats22/i154792

Changes related to comment discussion

victorsanni commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1234,6 +1235,14 @@ class CupertinoActionSheetAction extends StatefulWidget {
+  /// if `null`, cursor will default to:
+  /// - `MouseCursor.click` on web platforms
+  /// - `MouseCursor.defer` on other platforms
+  /// else will use value provided
```
https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#use-correct-grammar

victorsanni replied:
On a smaller note I'm wondering if there is any need for this comment change compared to the previous comment. Maybe the `if null` prefix?

srivats22 replied:
reason I changed the comment was mostly cause we are making the mouse cursor null. But I will rever to the below:

/// The cursor that will be shown when hovering over the button.
///
/// If not passed will default to [MouseCursor.defer] and on web will default to [SystemMouseCursors.click].

victorsanni replied:
'if not passed' and 'defaults' mean the same thing, so it would be redundant to have the 'if not passed' prefix. However, one can imagine a situation where a user passes `null` as the `mouseCursor`. So technically, in that case, something has been passed. But since we are null coalescing, the rest of the statement would apply.

Maybe a better comment would be `If null, defaults to...` Because that would cover the case where a mouse cursor is not passed and thus there is a default and the case where a null mouse cursor is passed.

victorsanni replied:
It seems like a subtle distinction, but it can make a huge difference for a developer trying to debug why a potentially null mouse cursor passed into this widget from somewhere else doesn't change the default to their expected cursor, because in reading the documentation they saw that the defaults are respected only if no mouse cursor is passed in.

srivats22 replied:
understood thank you for the comments I have modified it to the following:
/// The cursor that will be shown when hovering over the button.
///
/// if null defaults to [SystemMouseCursors.click] on web else [MouseCursor.defer].

victorsanni commented on packages/flutter/test/cupertino/action_sheet_test.dart:
```diff
@@ -1956,6 +1956,41 @@ void main() {
+                child: const Text('One'),
+              ),
+            ],
+          )
```
Missing a trailing comma.

victorsanni commented on packages/flutter/test/cupertino/action_sheet_test.dart:
```diff
@@ -1956,6 +1956,41 @@ void main() {
+
+    await tester.pumpWidget(
+      createAppWithButtonThatLaunchesActionSheet(
+          CupertinoActionSheet(
```
Outdent entire block by two spaces.

srivats22 replied:
<img width="1364" alt="Screenshot 2024-11-07 at 6 11 04 PM" src="https://github.com/user-attachments/assets/9ba9741f-04cf-4a95-9a72-c9ddbaf83c5a">
is this indentation ok? On my machine it seems to be indented properly

srivats22 replied:
<img width="1250" alt="Screenshot 2024-11-12 at 4 01 16 PM" src="https://github.com/user-attachments/assets/1dd42f40-7f95-4fd6-8ffe-80769741055a">
does this indentation look proper?

victorsanni commented on packages/flutter/test/cupertino/action_sheet_test.dart:
```diff
@@ -1956,6 +1956,41 @@ void main() {
+    await tester.pumpAndSettle();
+    expect(
+      RendererBinding.instance.mouseTracker.debugDeviceActiveCursor(1),
+      customCursor
```
Missing a trailing comma.

srivats22 committed:
Modified based on comments

srivats22 committed:
Merge pull request #5 from srivats22/i154792

Modified based on comments

victorsanni commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1234,6 +1235,11 @@ class CupertinoActionSheetAction extends StatefulWidget {
 
+  /// The cursor that will be shown when hovering over the button.
+  ///
+  /// if null defaults to [SystemMouseCursors.click] on web else [MouseCursor.defer].
```
```suggestion
  /// If null, defaults to [SystemMouseCursors.click] on web and [MouseCursor.defer] on other platforms.
```
https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#use-correct-grammar

victorsanni commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1312,7 +1318,7 @@ class _CupertinoActionSheetActionState extends State<CupertinoActionSheetAction>
 
     return MouseRegion(
-      cursor: kIsWeb ? SystemMouseCursors.click : MouseCursor.defer,
+      cursor: widget.mouseCursor ?? (kIsWeb ? SystemMouseCursors.click : MouseCursor.defer),
```
If a CupertinoActionSheetAction is disabled, shouldn't it not become clickable on web?

srivats22 replied:
I don't think it can be disabled because onClick is defined as required. So passing null or disabling isn't an option. I tried the same and it didn't work

srivats22 committed:
Update packages/flutter/lib/src/cupertino/dialog.dart

Co-authored-by: Victor Sanni <victorsanniay@gmail.com>

srivats22 committed:
Changing mouse cursor in test file

srivats22 committed:
Merge pull request #6 from srivats22/i154792

Changing mouse cursor in test file

victorsanni commented:
Fixes https://github.com/flutter/flutter/issues/154792

TahaTesser commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1234,6 +1235,11 @@ class CupertinoActionSheetAction extends StatefulWidget {
 
+  /// The cursor that will be shown when hovering over the button.
+  ///
+  /// If null, defaults to [SystemMouseCursors.click] on web and [MouseCursor.defer] on other platforms.
```
```suggestion
  /// If null, defaults to [SystemMouseCursors.click] on web and
  /// [MouseCursor.defer] on other platforms.
```

TahaTesser commented on packages/flutter/test/cupertino/action_sheet_test.dart:
```diff
@@ -1956,6 +1956,41 @@ void main() {
     );
   });
 
+  testWidgets('Hovering over Cupertino action sheet action updates cursor to passed cursor on any platform', (WidgetTester tester) async {
```
```suggestion
  testWidgets('CupertinoActionSheet action cursor behavior', (WidgetTester tester) async {
```

TahaTesser commented on packages/flutter/test/cupertino/action_sheet_test.dart:
```diff
@@ -1956,6 +1956,41 @@ void main() {
+    final TestGesture gesture = await tester.createGesture(kind: PointerDeviceKind.mouse, pointer: 1);
+    await gesture.addPointer(location: const Offset(10, 10));
+    await tester.pumpAndSettle();
+    expect(RendererBinding.instance.mouseTracker.debugDeviceActiveCursor(1), SystemMouseCursors.basic);
```
```suggestion
    expect(
      RendererBinding.instance.mouseTracker.debugDeviceActiveCursor(1),
      SystemMouseCursors.basic,
   );
```

TahaTesser reviewed (changes requested):
Thanks for the contribution!

srivats22 committed:
updated comment

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

srivats22 committed:
renaming test

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

srivats22 committed:
reformatting code

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

TahaTesser reviewed (approved):
LGTM
